File: apps/ios/Loonext/Features/Calls/Telephony/TelnyxSdkClient.swift
import Foundation
import CallKit
import AVFAudio
@preconcurrency import TelnyxRTC

/// The real `SoftphoneSdk` over the Telnyx iOS WebRTC SDK (TelnyxRTC 4.x).
/// One TxClient per registration; `connect` tears the old one down and builds
/// fresh — mint-on-connect is the whole recovery story (a dead token can't be
/// fixed by the SDK's own reconnect, a rebuild with a fresh mint can).
///
/// `@preconcurrency import`: the SDK predates strict concurrency; its
/// delegate calls back on its own threads, so every callback hops onto the
/// main actor before touching state.
@MainActor
final class TelnyxSdkClient: NSObject, SoftphoneSdk {
    var onEvent: ((SdkEvent) -> Void)?

    private var client: TxClient?
    private var handles: [UUID: TelnyxCallHandle] = [:]

    func connect(token: String, callerIdName: String, pushDeviceToken: String?) throws {
        disconnect()
        let next = TxClient()
        next.delegate = self
        client = next
        let config = TxConfig(
            token: token,
            // Registering the PushKit VoIP token at login is what lets the
            // Telnyx iOS push credential ring this device while the socket
            // is down. Nil until PushKit hands us one (simulator, or the
            // entitlement missing) — calls then ring only while connected.
            pushDeviceToken: pushDeviceToken,
            reconnectClient: true
        )
        try next.connect(txConfig: config)
    }

    func disconnect() {
        guard let previous = client else { return }
        client = nil
        // A stale client's straggler delegate events must not flip the fresh
        // one's state — cut the delegate before tearing it down.
        previous.delegate = nil
        previous.disconnect()
        handles.removeAll()
    }

    func newCall(
        callerIdName: String,
        callerIdNumber: String,
        destinationNumber: String,
        clientState: String
    ) throws -> any SdkCallHandle {
        guard let active = client else {
            throw ApiError(
                code: ApiErrorCode.network,
                message: "Calling isn't ready yet. Try again in a moment.",
                httpStatus: 0
            )
        }
        let callId = UUID()
        // The iOS SDK sends clientState to the wire AS-IS (see ClientState) —
        // this argument must already be the server's base64 value.
        let call = try active.newCall(
            callerName: callerIdName.isBlank ? "Loonext" : callerIdName,
            callerNumber: callerIdNumber,
            destinationNumber: destinationNumber,
            callId: callId,
            clientState: clientState
        )
        let handle = TelnyxCallHandle(call: call, uuid: callId)
        handles[callId] = handle
        return handle
    }

    func setAudioRoute(_ route: AudioRoute) {
        switch route {
        case .speaker:
            client?.setSpeaker()
        case .earpiece:
            client?.setEarpiece()
        }
    }

    // MARK: - CallKit passthroughs

    /// Answer via the SDK's CallKit path — it also parks the action until the
    /// pushed call's INVITE lands (the VoIP-wake answer race).
    func answerFromCallKit(_ action: CXAnswerCallAction) {
        client?.answerFromCallkit(answerAction: action)
    }

    /// End via the SDK's CallKit path (fulfills the action itself).
    func endFromCallKit(_ action: CXEndCallAction) {
        client?.endCallFromCallkit(endAction: action)
    }

    /// Whether the SDK currently tracks a call for this CallKit UUID.
    func hasCall(uuid: UUID) -> Bool {
        client?.getCall(callId: uuid) != nil
    }

    /// VoIP-push wake: hand the push metadata to the SDK so it connects and
    /// attaches the pushed INVITE. Creates the client when the app was
    /// launched cold by the push.
    func processVoIPNotification(
        token: String,
        pushDeviceToken: String?,
        pushMetaData: [String: Any]
    ) throws {
        let target: TxClient
        if let client {
            target = client
        } else {
            target = TxClient()
            target.delegate = self
            client = target
        }
        let config = TxConfig(
            token: token,
            pushDeviceToken: pushDeviceToken,
            reconnectClient: true
        )
        try target.processVoIPNotification(
            txConfig: config,
            serverConfiguration: TxServerConfiguration(),
            pushMetaData: pushMetaData
        )
    }

    /// MUST be driven from CXProviderDelegate didActivate/didDeactivate.
    func enableAudioSession(_ session: AVAudioSession) {
        client?.enableAudioSession(audioSession: session)
    }

    func disableAudioSession(_ session: AVAudioSession) {
        client?.disableAudioSession(audioSession: session)
    }

    // MARK: - Delegate plumbing (hops to MainActor)

    private func trackIncoming(_ call: TelnyxRTC.Call) {
        guard let info = call.callInfo else { return }
        let callId = info.callId
        if handles[callId] != nil { return }
        let handle = TelnyxCallHandle(call: call, uuid: callId)
        handles[callId] = handle
        onEvent?(.incoming(
            call: handle,
            callerName: info.callerName,
            callerNumber: info.callerNumber
        ))
    }

    private func routePhase(_ callId: UUID, _ phase: CallPhase?) {
        guard let handle = handles[callId] else { return }
        handle.onPhase?(phase)
        if phase == .ended {
            handles.removeValue(forKey: callId)
        }
    }

    static func phase(from state: CallState) -> CallPhase? {
        switch state {
        case .NEW, .CONNECTING, .RINGING:
            return .connecting
        case .ACTIVE:
            return .active
        case .HELD:
            return .held
        case .DONE:
            return .ended
        case .RECONNECTING, .DROPPED:
            // Mid-call recovery states — the call keeps its current phase; a
            // failed recovery lands in DONE on its own.
            return nil
        }
    }
}

extension TelnyxSdkClient: TxClientDelegate {
    nonisolated func onSocketConnected() {
        // Connected != registered; READY is the ring-eligible moment.
    }

    nonisolated func onSocketDisconnected() {
        Task { @MainActor in self.onEvent?(.disconnected) }
    }

    nonisolated func onClientError(error: Error) {
        let message = error.localizedDescription
        Task { @MainActor in self.onEvent?(.error(message)) }
    }

    nonisolated func onClientReady() {
        Task { @MainActor in self.onEvent?(.ready) }
    }

    nonisolated func onPushDisabled(success: Bool, message: String) {}

    nonisolated func onSessionUpdated(sessionId: String) {}

    nonisolated func onCallStateUpdated(callState: CallState, callId: UUID) {
        let mapped = Self.phase(from: callState)
        Task { @MainActor in self.routePhase(callId, mapped) }
    }

    nonisolated func onIncomingCall(call: TelnyxRTC.Call) {
        Task { @MainActor in self.trackIncoming(call) }
    }

    nonisolated func onRemoteCallEnded(callId: UUID, reason: CallTerminationReason?) {
        Task { @MainActor in self.routePhase(callId, .ended) }
    }

    /// A VoIP-push call's INVITE landed — same ring path as a live invite
    /// (the core dedups by call id).
    nonisolated func onPushCall(call: TelnyxRTC.Call) {
        Task { @MainActor in self.trackIncoming(call) }
    }
}

/// One SDK call behind the testable handle seam.
@MainActor
private final class TelnyxCallHandle: SdkCallHandle {
    let id: String
    let uuid: UUID
    private let call: TelnyxRTC.Call

    var onPhase: ((CallPhase?) -> Void)?

    init(call: TelnyxRTC.Call, uuid: UUID) {
        self.call = call
        self.uuid = uuid
        self.id = uuid.uuidString.lowercased()
    }

    var callControlId: String? { call.telnyxCallControlId }

    var telnyxSessionId: String? { call.telnyxSessionId?.uuidString.lowercased() }

    func answer() {
        call.answer()
    }

    func end() {
        call.hangup()
    }

    func hold() {
        call.hold()
    }

    func unhold() {
        call.unhold()
    }

    func setMuted(_ muted: Bool) {
        if muted {
            call.muteAudio()
        } else {
            call.unmuteAudio()
        }
    }

    func dtmf(_ digit: String) {
        call.dtmf(dtmf: digit)
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/telephony/TelnyxSdkClient.kt
package com.loonext.android.telephony

import android.content.Context
import com.telnyx.webrtc.sdk.TelnyxClient
import com.telnyx.webrtc.sdk.TokenConfig
import com.telnyx.webrtc.sdk.model.AudioDevice
import com.telnyx.webrtc.sdk.model.CallState
import com.telnyx.webrtc.sdk.model.LogLevel
import com.telnyx.webrtc.sdk.model.SocketMethod
import com.telnyx.webrtc.sdk.model.SocketStatus
import com.telnyx.webrtc.sdk.model.TxServerConfiguration
import com.telnyx.webrtc.sdk.verto.receive.InviteResponse
import com.telnyx.webrtc.sdk.verto.receive.ReceivedMessageBody
import com.telnyx.webrtc.sdk.verto.receive.SocketResponse
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

/**
 * The real [SdkClient] over the Telnyx Android WebRTC SDK (v3.5). One
 * TelnyxClient per registration; [connect] tears the old one down and builds
 * fresh — mint-on-connect is the whole recovery story (a dead token can't be
 * fixed by the SDK's own reconnect, a rebuild with a fresh mint can).
 */
class TelnyxSdkClient(
    context: Context,
    private val scope: CoroutineScope,
) : SdkClient {
    private val appContext = context.applicationContext

    private val _events = MutableSharedFlow<SdkEvent>(extraBufferCapacity = 32)
    override val events: SharedFlow<SdkEvent> = _events

    private var client: TelnyxClient? = null
    private var socketJob: Job? = null

    override fun connect(token: String, callerIdName: String) {
        disconnect()
        val next = TelnyxClient(appContext)
        client = next
        socketJob = scope.launch {
            next.socketResponseFlow.collect { response ->
                // #168A: one malformed/unexpected message must not kill this
                // collector (every future event) — or, uncaught, the process.
                runCatching { onSocketResponse(next, response) }
            }
        }
        next.connect(
            TxServerConfiguration(),
            TokenConfig(
                sipToken = token,
                sipCallerIDName = callerIdName.ifBlank { "Loonext" },
                sipCallerIDNumber = "",
                fcmToken = null,
                ringtone = null,
                ringBackTone = null,
                logLevel = LogLevel.ERROR,
                autoReconnect = true,
            ),
            null,
            true,
        )
    }

    override fun disconnect() {
        socketJob?.cancel()
        socketJob = null
        val previous = client
        client = null
        runCatching { previous?.disconnect() }
    }

    override fun newCall(
        callerIdName: String,
        callerIdNumber: String,
        destinationNumber: String,
        clientState: String,
    ): SdkCallHandle {
        val active = client ?: throw IllegalStateException("softphone is not connected")
        // The Android SDK base64-encodes its clientState argument internally;
        // hand it the DECODED tag so the wire carries the server's client_state
        // byte-for-byte (see ClientState).
        val call = active.newInvite(
            callerIdName,
            callerIdNumber,
            destinationNumber,
            ClientState.forAndroidSdk(clientState),
        )
        return TelnyxCallHandle(active, call)
    }

    override fun setAudioRoute(route: AudioRoute) {
        client?.setAudioOutputDevice(
            when (route) {
                AudioRoute.EARPIECE -> AudioDevice.PHONE_EARPIECE
                AudioRoute.SPEAKER -> AudioDevice.LOUDSPEAKER
                AudioRoute.BLUETOOTH -> AudioDevice.BLUETOOTH
            },
        )
    }

    private fun onSocketResponse(
        source: TelnyxClient,
        response: SocketResponse<ReceivedMessageBody>,
    ) {
        // A stale client's straggler events must not flip the fresh one's state.
        if (source !== client) return
        when (response.status) {
            SocketStatus.ESTABLISHED, SocketStatus.LOADING -> Unit
            SocketStatus.DISCONNECT -> _events.tryEmit(SdkEvent.Disconnected)
            SocketStatus.ERROR -> _events.tryEmit(SdkEvent.Error(response.errorMessage))
            SocketStatus.MESSAGERECEIVED -> onMessage(source, response.data)
        }
    }

    private fun onMessage(source: TelnyxClient, body: ReceivedMessageBody?) {
        when (body?.method) {
            SocketMethod.CLIENT_READY.methodName -> _events.tryEmit(SdkEvent.Ready)
            SocketMethod.INVITE.methodName -> {
                val invite = body.result as? InviteResponse ?: return
                val call = source.getActiveCalls()[invite.callId] ?: return
                // §3.2 DETERMINISTIC correlation: the server stamps
                // `X-Loonext-Session` on every member ring dial, and the SDK
                // surfaces it here as first-class custom headers ({name,value},
                // verified against telnyx-webrtc-android v3.5.0). Hand the raw
                // list up; [TelecomCallReducer.correlateInvite] reads the header
                // — NEVER a caller/time guess. legId is the by-leg fallback key
                // for an older server that hasn't shipped the header yet.
                val headers = runCatching {
                    invite.customHeaders?.mapNotNull { header ->
                        val name = header.name ?: return@mapNotNull null
                        name to (header.value ?: "")
                    }.orEmpty()
                }.getOrDefault(emptyList())
                val legId = runCatching { call.getTelnyxSessionId()?.toString() }.getOrNull()
                _events.tryEmit(
                    SdkEvent.Incoming(
                        call = TelnyxCallHandle(source, call),
                        callerName = invite.callerIdName,
                        callerNumber = invite.callerIdNumber,
                        customHeaders = headers,
                        legId = legId,
                    ),
                )
            }
        }
    }
}

/** One SDK call behind the testable handle interface. */
private class TelnyxCallHandle(
    private val client: TelnyxClient,
    private val call: com.telnyx.webrtc.sdk.Call,
) : SdkCallHandle {
    override val id: String = call.callId.toString()

    override val callControlId: String?
        get() = call.getTelnyxCallControlId()

    override val telnyxSessionId: String?
        get() = call.getTelnyxSessionId()?.toString()

    override val phases: Flow<CallPhase?> = call.callStateFlow.map { state ->
        when (state) {
            is CallState.NEW, is CallState.CONNECTING, is CallState.RINGING ->
                CallPhase.CONNECTING

            is CallState.ACTIVE -> CallPhase.ACTIVE
            is CallState.HELD -> CallPhase.HELD
            is CallState.DONE, is CallState.ERROR -> CallPhase.ENDED

            // Mid-call recovery states — the call keeps its current phase; a
            // failed recovery lands in DONE/ERROR on its own.
            else -> null
        }
    }

    override fun accept(destinationNumber: String) {
        client.acceptCall(call.callId, destinationNumber)
    }

    override fun end() {
        client.endCall(call.callId)
    }

    override fun toggleHold() {
        call.onHoldUnholdPressed(call.callId)
    }

    override fun setMuted(muted: Boolean) {
        call.setMuteState(muted)
    }

    override fun dtmf(digit: String) {
        call.dtmf(call.callId, digit)
    }
}
